Godeps/_workspace/src/github.com/jbenet/go-reuseport/impl_unix.go

```diff
@@ -0,0 +1,339 @@
+		return -1, err
+	}
+
+	if protocol == syscall.IPPROTO_TCP {
+		//  by default golang/net sets TCP no delay to true.
+		if err = setNoDelay(fd, true); err != nil {
+			// fmt.Println("set no delay failed")
+			syscall.Close(fd)
+			return -1, err
+		}
+	}
+
+	if err = syscall.SetNonblock(fd, true); err != nil {
+		syscall.Close(fd)
+		return -1, err
+	}
+
+	return fd, nil
+}
+
+func listenStream(netw, addr string) (l net.Listener, err error) {
+	var (
+		file *os.File
+	)
+
+	fd, err := listen(netw, addr)
+	if err != nil {
+		return nil, err
+	}
+
+	// Set backlog size to the maximum
+	if err = syscall.Listen(fd, syscall.SOMAXCONN); err != nil {
+		// fmt.Println("listen failed")
+		syscall.Close(fd)
+		return nil, err
+	}
+
+	file = os.NewFile(uintptr(fd), filePrefix+strconv.Itoa(os.Getpid()))
+	if l, err = net.FileListener(file); err != nil {
+		// fmt.Println("filelistener failed")
+		syscall.Close(fd)
+		return nil, err
+	}
+
+	if err = file.Close(); err != nil {
+		// fmt.Println("file close failed")
+		syscall.Close(fd)
+		return nil, err
+	}
+
+	return l, err
+}
+
+func listenPacket(netw, addr string) (p net.PacketConn, err error) {
+	var (
+		file *os.File
+	)
+
+	fd, err := listen(netw, addr)
+	if err != nil {
+		return nil, err
+	}
+
+	file = os.NewFile(uintptr(fd), filePrefix+strconv.Itoa(os.Getpid()))
+	if p, err = net.FilePacketConn(file); err != nil {
+		// fmt.Println("filelistener failed")
+		syscall.Close(fd)
+		return nil, err
+	}
+
+	if err = file.Close(); err != nil {
+		// fmt.Println("file close failed")
+		syscall.Close(fd)
+		return nil, err
+	}
+
+	return p, err
+}
+
+func listenUDP(netw, addr string) (c net.Conn, err error) {
+	var (
+		file *os.File
+	)
+
+	fd, err := listen(netw, addr)
+	if err != nil {
+		return nil, err
+	}
+
+	file = os.NewFile(uintptr(fd), filePrefix+strconv.Itoa(os.Getpid()))
+	if c, err = net.FileConn(file); err != nil {
+		// fmt.Println("filelistener failed")
+		syscall.Close(fd)
+		return nil, err
+	}
+
+	if err = file.Close(); err != nil {
+		// fmt.Println("file close failed")
+		syscall.Close(fd)
+		return nil, err
+	}
+
+	return c, err
+}
+
+func connect(fd int, ra syscall.Sockaddr) error {
+	switch err := syscall.Connect(fd, ra); err {
+	case syscall.EINPROGRESS, syscall.EALREADY, syscall.EINTR:
+	case nil, syscall.EISCONN:
+		return nil
+	default:
+		return err
+	}
+
+	var err error
+	for {
+		// if err := fd.pd.WaitWrite(); err != nil {
+		// 	return err
+		// }
+		// i'd use the above fd.pd.WaitWrite to poll io correctly, just like net sockets...
+		// but of course, it uses fucking runtime_* functions that _cannot_ be used by
+		// non-go-stdlib source... seriously guys, what kind of bullshit is that!?
+		<-time.After(20 * time.Microsecond)
+		var nerr int
+		nerr, err = syscall.GetsockoptInt(fd, syscall.SOL_SOCKET, syscall.SO_ERROR)
+		if err != nil {
+			return err
+		}
+		switch err = syscall.Errno(nerr); err {
+		case syscall.EINPROGRESS, syscall.EALREADY, syscall.EINTR:
```

Comment:
yeah. good instinct. in this case looks like `EAGAIN` on `connect()` in linux is a larger connect problem: http://man7.org/linux/man-pages/man2/connect.2.html#EAGAIN -- not the usual R/W `EAGAIN`